File: Python/ByTier/A/0963_A_Alternating_Sum.py
# Contest 963, Problem A: Alternating Sum
# URL: https://codeforces.com/contest/963/problem/A

M=10**9+9
n,a,b,k=[int(e) for e in input().split()]
seq=[1 if e=='+' else -1 for e in input()]
s=0
x=pow(a,n,M)
y=(pow(a,-1,M)*b)%M
for i in range(k):
    s+=x*seq[i]
    x*=y
    s%=M
    x%=M
q=pow(y,k,M)
S=s*((n+1)//k)
if q!=1:
    S=s*(pow(q,(n+1)//k,M)-1)*pow(q-1,-1,M)
print(S%M)

File: CPP/ByTier/A/0963_A_Alternating_Sum.cpp
/*
 * Problem URL : https://codeforces.com/contest/963/problem/A
 * Submit Date : 2025-09-04
 */

#include <bits/stdc++.h>
using namespace std;
#define ll long long
const ll md = 1e9 + 9;
ll n, a, b, k, ans, sum, xa, xi, x, ia;
ll ksm(ll a, ll b) {
  ll re = 1;
  a %= md;
  while (b) {
    if (b & 1) {
      re = re * a % md;
    }
    a = a * a % md;
    b /= 2;
  }
  return re;
}
string s;
int main() {
  std::ios::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);
  cin >> n >> a >> b >> k;
  cin >> s;
  ans = 0;
  x = ksm(a, n);
  ia = ksm(a, md - 2), xa = 1;
  for (int i = 0; i < k; i++) {
    if (s[i] == '+') {
      ans = (ans + x * xa % md) % md;
    } else {
      ans = (ans + md - x * xa % md) % md;
    }
    x = x * ia % md, xa = xa * b % md;
  }
  xa = 1;
  sum = 0;
  xi = ksm(b * ia, k);
  x = (n + 1) / k;
  if (xi == 1) {
    cout << ans * x % md << endl;
  } else {
    ll xxx = (1 + md - ksm(xi, x));
    ll xxx1 = ksm(1 + md - xi, md - 2);
    cout << ans * xxx % md * xxx1 % md << endl;
  }
}